### PR TITLE
Kafka retry

### PR DESCRIPTION
**Pull request by mtps**
* README pending
See example usage [here](https://github.com/provenance-io/kafka-coroutines/compare/mtps/kafka-retry?expand=1#diff-f18aab011df807c33ad63bb8883686fac30e778c086efc67f09f071ad7c9bb87)

### CODE REVIEW

File: kafka-coroutines-retry/src/main/kotlin/io/provenance/kafka/coroutines/retry/KafkaErrorHandler.kt [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        }
+    }
+}
+
+/**
+ * Lift the lambda callback from [UnAckedConsumerRecord] to [ConsumerRecord] type.
+ *
+ * This is needed to match the flow type of [tryOnEach] when processing with [KafkaFlowRetry].
+ */
+fun <K, V> (suspend (ConsumerRecord<K, V>) -> Unit).lifted(): suspend (UnAckedConsumerRecord<K, V>) -> Unit =
```
Comment (piercetrey-figure) [inline kafka-coroutines-retry/src/main/kotlin/io/provenance/kafka/coroutines/retry/KafkaErrorHandler.kt:37]: maybe a dumb question, but when you convert the UnAckedConsumerRecord to a ConsumerRecord, what happens with the ack?

File: kafka-coroutines-retry/src/main/kotlin/io/provenance/kafka/coroutines/retry/flow/RetryFlow.kt [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    flowRetry: FlowRetry<T>,
+    retryInterval: Duration = DEFAULT_RETRY_INTERVAL,
+    retryStrategies: List<RetryStrategy> = defaultRetryStrategies
+): Flow<T> {
+    val log = KotlinLogging.logger {}
+    val strategies = retryStrategies.invert()
+
+    return pollingFlow(retryInterval) {
+        for (strategy in strategies) {
+            val lastAttempted = OffsetDateTime.now().minus(strategy.value.lastAttempted.toJavaDuration())
```
Comment (rchaing-figure) [inline kafka-coroutines-retry/src/main/kotlin/io/provenance/kafka/coroutines/retry/flow/RetryFlow.kt:45]: im dumb, how does retry strategy work?

**Review by rchaing-figure — APPROVED** [review]
just one question

**Review by piercetrey-figure — APPROVED** [review]
looks pretty nifty

File: kafka-coroutines-retry/src/main/kotlin/io/provenance/kafka/coroutines/retry/flow/RetryFlow.kt (hunk 2 above)
Comment (mtps) [reply kafka-coroutines-retry/src/main/kotlin/io/provenance/kafka/coroutines/retry/flow/RetryFlow.kt:45]: for example:
`RetryStrategy(5, 10.minutes), RetryStrategy(4, 1.days)` would retry 5 times, with 10 minutes in between; once those are exhausted, retry 4 times, with 1 day in between. it's a defined back-off strategy.

File: kafka-coroutines-retry/src/main/kotlin/io/provenance/kafka/coroutines/retry/KafkaErrorHandler.kt (hunk 1 above)
Comment (mtps) [reply kafka-coroutines-retry/src/main/kotlin/io/provenance/kafka/coroutines/retry/KafkaErrorHandler.kt:37]: Nothing, it is maintained in the underlying flow (allowing the `acking()` flow call after the fact). the conversion happens in a `tryOnEach()` (which mirrors the standard kotlin flow's `onEach` method.

Comment (piercetrey-figure) [reply kafka-coroutines-retry/src/main/kotlin/io/provenance/kafka/coroutines/retry/KafkaErrorHandler.kt:37]: got it, thanks for the info!

**Pull request merged**